Allow Multiple Logic Requests

GCRA101 opened the pull request:
### Issues addressed by this PR

 Closes #465 

![GH Script Canvas View](https://github.com/user-attachments/assets/8d36d9d2-47ea-40ae-901c-166aea21297a)


ETABS Toolkit now can allow to pull objects from ETABS based on FilterRequests, SelectedRequest and multiple LogicalRequests.


 ### Test files
Grasshopper File
https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/ETABS_Toolkit/%23474-AllowMultipleLogicRequests/[TestScript.gh](https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/ETABS_Toolkit/%23474-AllowMultipleLogicRequests/TestScript.gh?csf=1&web=1&e=93kD0Q)?csf=1&web=1&e=93kD0Q
ETABS File
https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/ETABS_Toolkit/%23474-AllowMultipleLogicRequests/Test%20ETABS%20Model.EDB?csf=1&web=1&e=IRUEie

 ### Changelog

- Added Read Method for FitlerRequests
- Added Read Method for LogicalRequests (And, Or and Not)
- Added Dynamic Comparer class allowing to compare IBHoMObjects based on ad-hoc criteria

GCRA101 committed:
Add Read Method for pulling IRequests

GCRA101 committed:
Add Read Method for both FilterRequests and ILogicalRequests

GCRA101 committed:
Add Reflection-based Dynamic Comparer class to allow to intersect Hashsets of IBHoMObjects

GCRA101 committed:
Fix bug

GCRA101 committed:
Turn HashSet into List to allow for use of .Equals method from Dynamic Comparer

GCRA101 committed:
Fix minor issues

GCRA101 committed:
Add code to read from LogicalNotRequests

GCRA101 committed:
Add comments highlighting use of GENERIC TYPES, HASH TABLES, STREAMS, RECURSION AND REFLECTION

GCRA101 committed:
Remove outstanding lines of code

GCRA101 committed:
Improve Hash function for the DynamicCompararer

GCRA101 committed:
Improve the Dynamic Comparer methods .Equals() and .GetHashCode() and adjust recursion calls

GCRA101 committed:
Add comments

GCRA101 commented:
@ChrisBuckle , @peterjamesnugent , @IsakNaslundBh,

I've had the chance to finish the pull request for issue #465.

Also for this one, a pull request to be reviewed in the next sprint moving forward BHoM 7.4, not now ;)

Thanks guys! ;)

Chrisshort92 reviewed (dismissed):
Functionality tested with files provided, all working as expected. Approved for merge

Chrisshort92 commented:
@BHoMBot Check required

GCRA101 committed:
Update _Read.cs

Turned the getEtabsId() method name (from the DynamicComparer private class) into GetEtabsId(), based on BHoM method's naming convention.

GCRA101 committed:
Turn DynamicComparer private class into public class in Types namespace

GCRA101 commented:
@BHoMBot Check required

Chrisshort92 reviewed (dismissed):
Re-tested code changes from @GCRA101, functionality working as expected, approved for merge.

peterjamesnugent commented:
@BHoMBot check compliance
@BHoMBot check ready-to-merge

GCRA101 committed:
Add copyright header to DynamicComparer class

GCRA101 commented:
@peterjamesnugent, 
added copyright header to DynamicComparer in latest commit.
9b081de06fd93e990711bee8fd65c5e748cb3113

GCRA101 commented:
@BHoMBot check required

GCRA101 commented:
@BHoMBot check copyright-compliance

GCRA101 committed:
Repaste Copyright header in DynamicComparer

peterjamesnugent commented on Etabs_Adapter/Convert/Execute.cs:
```diff
@@ -0,0 +1,233 @@
+        /**** IAdapter Interface                        ****/
+        /***************************************************/
+
+        public override Output<List<object>, bool> Execute(IExecuteCommand command, ActionConfig actionConfig = null)
+        {
+            var output = new Output<List<object>, bool>() { Item1 = null, Item2 = false };
+
+            output.Item2 = RunCommand(command as dynamic);
+
+            return output;
+        }
```
Have we actually made any changes to this file? Doesn't look like it.

I think you've moved a file by accident.

peterjamesnugent commented on Etabs_Adapter/Types/DynamicComparer.cs:
```diff
@@ -0,0 +1,103 @@
+
+                Type objType = obj1.GetType();
+
+                // For physical objects, use Id as it is never null
+                if (objType == typeof(Node) || objType == typeof(Bar) || objType == typeof(Panel) || objType == typeof(RigidLink) || objType == typeof(FEMesh))
+                    return GetEtabsId(obj1).Id.ToString() == GetEtabsId(obj2).Id.ToString();
+                // For all other items, use Name as it is never null
+                if (objType == typeof(ISectionProperty) || objType == typeof(IMaterialFragment) || objType == typeof(ISurfaceProperty) ||
+                    objType == typeof(Loadcase) || objType == typeof(LoadCombination) || objType == typeof(ILoad) || objType == typeof(LinkConstraint) ||
+                    objType == typeof(Level) || objType == typeof(Grid))
+                    return obj1.Name.ToString() == obj2.Name.ToString();
```
We already have all these comparers defined in the Adapter - can we not make sure of that rather than rewriting it here:

https://github.com/BHoM/ETABS_Toolkit/blob/edc11e02ca0a6e0b06a6a9a721e7c98a9a380295/Etabs_Adapter/Types/Comparer.cs#L57-L62

GCRA101 replied:
@peterjamesnugent, thanks for the comment.
It would be great to reuse it yeah.
Only thing is that there are some objects that are not dealt with by the Adapter Comparer (e.g. Level, FEMesh,LoadCase...) - see screenshot below.
Can we add them in it?

![image](https://github.com/user-attachments/assets/a6aa5d4b-847d-42da-9edd-701b7a8f5a39)

peterjamesnugent replied:
Yes, I don't think that's an issue adding them to the `Comparer.cs` - most of them will be `NameOrDescriptionComparer()` anyway.

Can you add to the test script, trying to push two objects with the same name for each type you add (just to test that we are not breaking anything).

peterjamesnugent reviewed (changes requested):
Recurring theme of unchanged files popping up in the PRs, can we make sure they are removed.

On the `DynamicComparer` I feel like we can just use the comparers defined in the adatper rather than trying to rewrite those.

GCRA101 committed:
Revert Execute.cs to original

Chrisshort92 commented:
@BHoMBot check installer
